cc-31.1/man/size.1
.\"	@(#)size.1	6.1 (Berkeley) 4/29/85
.\"
.TH SIZE 1 "June 14, 1989"
.AT 3
.SH NAME
size \- size of an object file
.SH SYNOPSIS
.B size 
[ option ] [ object ... ]
.SH DESCRIPTION
.I Size
(without the
.B \-m
option) prints the (decimal) number of bytes required by the \_\^\_TEXT,
\_\^\_DATA and \_\^\_OBJC segments.  All other segments are totaled and
that size is listed in the `others' column.  The final two columns is 
the sum in hex and decimal.
If no file is specified,
.B a.out
is used.
.PP
There is one option to
.IR size (1):
.TP
.B \-m
Print the sizes of the Mach-O segments and sections as well as the total sizes
of the sections in each segment and the total size of the segments in the file.
.SH "SEE ALSO"
otool(1)
.SH BUGS
The size of common symbols can't be reflected in any of the numbers for
relocatable object files.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cc-31.1/misc/size.c
static	char *sccsid = "@(#)size.c	4.4 (Berkeley) 4/22/85";
/*
 * size
 */

#include <stdio.h>
#include <a.out.h>
#include <sys/loader.h>
#define PAGESIZE 8192

int header = 0;
int nfiles = 0;
int err = 0;
int mflag = 0;
int lflag = 0;

void
main(
    int argc,
    char **argv,
    char **envp)
{
	int i, args_left;

	for (i = 1; i < argc; i++) {
	    if(argv[i][0] == '-'){
		if(argv[i][1] == '\0'){
		    nfiles += argc - i - 1;
		    break;
		}
		if(strcmp(argv[i], "-m") == 0){
		    mflag = 1;
		    continue;
		}
		if(strcmp(argv[i], "-l") == 0){
		    lflag = 1;
		    continue;
		}
	    }
	    nfiles++;
	}

	args_left = 1;
	for (i = 1; i < argc; i++) {
	    if(args_left && argv[i][0] == '-'){
		if(argv[i][1] == '\0'){
		    args_left = 0;
		    continue;
		}
		if(strcmp(argv[i], "-m") == 0)
		    continue;
		if(strcmp(argv[i], "-l") == 0)
		    continue;
	    }
	    size(argv[i]);
	}
	if(nfiles == 0)
	    size("a.out");
	exit(err);
}

size(
    char *name)
{
	struct exec buf;
	long text, data, objc, others, sum;
	int i, j;
	FILE *f;

	int mach;
	struct mach_header mh;
	struct load_command *load_commands, *lcp;
	struct segment_command *sgp;
	struct section *sp;
	long seg_sum, sect_sum;

	if((f = fopen(name, "r")) == NULL){
	    printf("size: %s not found\n", name);
	    err++;
	    return;
	}
	if(fread((char *)&buf, sizeof(buf), 1, f) != 1){
	    printf("size: can't read header of %s\n", name);
	    err++;
	    return;
	}
	mach = *((long *)&buf) == MH_MAGIC;
	if(mach){
	    rewind(f);
	    if(fread((char *)&mh, sizeof(struct mach_header), 1, f) != 1){
		printf("size: can't read header of %s\n", name);
		err++;
		return;
	    }
	    load_commands = (struct load_command *)malloc(mh.sizeofcmds);
	    if(load_commands == NULL){
		printf("size: out of memory\n");
		exit(1);
	    }
	    if(fread((char *)load_commands, mh.sizeofcmds, 1, f) != 1){
		printf("size: can't read load commands of %s\n", name);
		err++;
		return;
	    }
	    if(mflag){
		if(nfiles > 1)
			printf("%s:\n", name);
		lcp = load_commands;
		seg_sum = 0;
		for(i = 0; i < mh.ncmds; i++){
		    if(lcp->cmd == LC_SEGMENT){
			sgp = (struct segment_command *)lcp;
			printf("Segment %0.16s: %d", sgp->segname, sgp->vmsize);
			if(sgp->flags & SG_FVMLIB)
			    printf(" (fixed vm library segment)\n");
			else
			    printf("\n");
			seg_sum += sgp->vmsize;
			sp = (struct section *)((char *)sgp +
				sizeof(struct segment_command));
			sect_sum = 0;
			for(j = 0; j < sgp->nsects; j++){
			    if(mh.filetype == MH_OBJECT)
				printf("\tSection (%0.16s, %0.16s): %d",
				       sp->segname, sp->sectname, sp->size);
			    else
				printf("\tSection %0.16s: %d", sp->sectname,
				       sp->size);
			    if(lflag)
				printf(" (addr 0x%x offset %d)\n",
					sp->addr, sp->offset);
			    else
				printf("\n");
			    sect_sum += sp->size;
			    sp++;
			}
			if(sgp->nsects > 0)
			    printf("\ttotal %d\n", sect_sum);
		    }
		    lcp = (struct load_command *)((char *)lcp + lcp->cmdsize);
		}
		printf("total %d\n", seg_sum);
		free(load_commands);
		return;
	    }
	    text = 0;
	    data = 0;
	    objc = 0;
	    others = 0;
	    lcp = load_commands;
	    for(i = 0; i < mh.ncmds; i++){
		if(lcp->cmd == LC_SEGMENT){
		    sgp = (struct segment_command *)lcp;
		    if(mh.filetype == MH_OBJECT){
			sp = (struct section *)((char *)sgp +
				sizeof(struct segment_command));
			for(j = 0; j < sgp->nsects; j++){
			    if(strcmp(sp->segname, SEG_TEXT) == 0)
				text += sp->size;
			    else if(strcmp(sp->segname, SEG_DATA) == 0)
				data += sp->size;
			    else if(strcmp(sp->segname, SEG_OBJC) == 0)
				objc += sp->size;
			    else
				others += sp->size;
			    sp++;
			}
		    }
		    else{
			if(strcmp(sgp->segname, SEG_TEXT) == 0)
			    text += sgp->vmsize;
			else if(strcmp(sgp->segname, SEG_DATA) == 0)
			    data += sgp->vmsize;
			else if(strcmp(sgp->segname, SEG_OBJC) == 0)
			    objc += sgp->vmsize;
			else
			    others += sgp->vmsize;
		    }
		}
		lcp = (struct load_command *)((char *)lcp + lcp->cmdsize);
	    }
	    free(load_commands);
	}
	else if(N_BADMAG(buf) && !mach){
	    printf("size: %s not an object file\n", name);
	    fclose(f);
	    err++;
	    return;
	}
	else{
	    text = buf.a_text;
	    data = buf.a_data + buf.a_bss;
	    if(buf.a_magic == ZMAGIC)
	        data = (data + PAGESIZE - 1) & ~(PAGESIZE - 1);
	    objc = 0;
	    others = 0;
	}
	if(header == 0){
	    printf("__TEXT\t__DATA\t__OBJC\tothers\tdec\thex\n");
	    header = 1;
	}
	printf("%u\t%u\t%u\t%u\t", text, data, objc, others);
	sum = text + data + objc + others;
	printf("%ld\t%lx", sum, sum);
	if(nfiles > 1)
	    printf("\t%s", name);
	printf("\n");
	fclose(f);
}
